Configurations/unix-Makefile.tmpl

```diff
@@ -853,13 +853,13 @@ tags TAGS: FORCE
 # Release targets (note: only available on Unix) #####################
 
 tar:
-	(cd $(SRCDIR); ./util/mktar.sh --name='$(NAME)' --tarfile='$(TARFILE)')
+	( cd "$(SRCDIR)"; ./util/mktar.sh --name='$(NAME)' --tarfile='$(TARFILE)' )
 
 # Helper targets #####################################################
 
-link-utils: $(BLDDIR)/util/opensslwrap.sh
+link-utils: $(subst $(subst x, ,x),\ ,$(BLDDIR)/util/opensslwrap.sh)
```

Comment:
This could probably better be in a function?
BTW: What is the purpose of this?

Reply:
This is not OK either way, as it requires GNU make.

Reply:
Oh, in that case, I need to be passing on this, sorry. I wasn't aware that there is any other kind of (popular?) make on \*nix than the GNU one. At least the problem spots are idendified now, so it can be taken from there, depending on requirements/etc.

Reply:
I don't think we need to support overriding BLDDIR and/or SRCDIR
without executing Configure.

Reply:
> I wasn't aware that there is any other kind of (popular?) make on *nix than the GNU one

BSD have their own.  It is *different*.  I believe that most commercial Unix vendors have their own, which might or might not extend the classic old (and dumb) make.

Reply:
Short of experience with these, I can't help here, but the simple `subst()` logic can easily be generated differently depending on the target `make` tool, depending of course on how space-escaping needs to be done on these make platforms. For now it may be a viable solution to limit the scope of `subst()` to GNU make (how to filter for that in the Perl generator?) and extend that later with other `make` flavours. The only thing "lost" here is that these make tools will _remain_ broken when spaces are involved (likely not a huge issue overall). Which makes this PR still worthy to consider.

Another, possibly more generic solution is to try avoiding variables in rules, but I suspect that at the end, some make-specific syntax will still need to be employed to support `SRCDIR`/`BLDDIR` (such as `VPATH`). Not knowing these other make flavours, it's hard to say what is more portable/easier (though BSD Make does support it as well.)

Reply:
Well, in this particular case, I don't think you need to worry, 'cause `$(BLDDIR)` should always be '.'.

However, I *think* that the common trick to have spaces in file names that `make` can deal with is this:

``` make
EMPTY=
SPACE=$(EMPTY) $(EMPTY)

...

whatever: my$(SPACE)spacy$(SPACE)file
```

And then, to get source directory specs right, in case they have spaces, could done with perl:

```
whatever: {- (my $x = $config{sourcedir}) =~ s| |\$(SPACE)|g; $x -}/some/thing
```

Note: *I haven't tested this*...  I just vaguely remember something like that from a text I read ages ago.

Reply:
Note, however that when I played with this PR, I think I noticed,
that some dependencies used $(SRCDIR) and some were quoting the SRCDIR as
it was at configure time, so changing SRCDIR after configure is finished will
likely not work at all.

Reply:
So maybe we should do this:

``` make
SRCDIR={- (my $x = $config{sourcedir}) =~ s| |\$(SPACE)|g; $x -}
```

or well, we could have this in the starting perl snippet and use that:

``` perl
    (our $srcdir = $config{sourcedir}) =~ s| |\$(SPACE)|g;
```

and then have:

``` make
SRCDIR={- $srcdir -}
```

Reply:
Neat trick and (on GNU Make) at least, it works as expected with a backslash added to escape the space (without it, it breaks the value into two):
```make
EMPTY=
SPACE=$(EMPTY)\ $(EMPTY)

VALUE=az$(SPACE)c

az$(SPACE)b:
	@echo

$(VALUE):
	@echo

test1: az$(SPACE)b
	echo 'test1'

test2: $(VALUE)
	echo 'test2'
```

```console
$ make -d test1 test2 | grep 'az'                                                                                                                                      (git)-[master] 
  Considering target file `az b'.
   File `az b' does not exist.
   Finished prerequisites of target file `az b'.
  Must remake target `az b'.
Putting child 0x7ff64dc0ff50 (az b) PID 42145 on the chain.
Live child 0x7ff64dc0ff50 (az b) PID 42145 
  Successfully remade target file `az b'.
  Considering target file `az c'.
   File `az c' does not exist.
   Finished prerequisites of target file `az c'.
  Must remake target `az c'.
Putting child 0x7ff64de00690 (az c) PID 42159 on the chain.
Live child 0x7ff64de00690 (az c) PID 42159 
  Successfully remade target file `az c'.
```

But we still need to original, unescaped value for all other non-make-rule purposes (they may confuse normal make/shell/perl code, so we must introduce a hard-wired variation of `SRCDIR`/`BLDDIR` values with the above trick applied, and use that in rule definitions:

```make
EMPTY=
SPACE=$(EMPTY)\ $(EMPTY)
SRCDIR=az b
BLDDIR=az c
SRCDIRINRULE=az$(SPACE)b
BLDDIRINRULE=az$(SPACE)c

[...]
link-utils: $(BLDDIRINRULE)/util/opensslwrap.sh
[...]

$(BLDDIRINRULE)/util/opensslwrap.sh: configdata.pm
[...]

configdata.pm: $(SRCDIRINRULE)/Configure $(SRCDIRINRULE)/config [...]
```

What do you think?

Reply:
Why do we need a separate SRCDIRINRULE / BLDDIRINRULE?  Where will the escaped space confuse?

Reply:
In bash the backslash will remain there.

Reply:
Meh, I thought bash expanded all backslashes in strings...  apparently not.

Anyhow, backslashes like you tried do not work universally:

foo.mk:
``` make
EMPTY=
SPACE=$(EMPTY)\ $(EMPTY)
FOO=ab$(SPACE)cb

all: $(FOO)
	echo "$(FOO)"
```

``` console
: ; bmake -f foo.mk
bmake: don't know how to make ab\. Stop

bmake: stopped in /home/levitte/tmp
```

(`bmake` is BSD make, packaged on Debian)

I do have a vague recollection about going through this already, a few years ago.

Reply:
Is there any way at all to use space in rule filenames with BSD Make? By trial and error I couldn't find any.